### PR TITLE
 timeout added to readiness probe

### PR DESCRIPTION
**Pull request by rmpt**

### CODE REVIEW

File: src/main/kotlin/test/ktortemplate/Application.kt [hunk 1]
```diff
@@ -76,8 +77,10 @@ fun Application.module(configOverrides: ApplicationConfig? = null) {
     }
 
     install(Health) {
+        val readinessDbTimeout: ApplicationConfigValue? = environment.config
```
Comment (felix19350) [inline src/main/kotlin/test/ktortemplate/Application.kt:80]: you may be able to simply declare 
val readinessDbTimeout = environment.config.propertyOrNull("healthcheck.readiness.database.timeoutMillis")

The type system should be able to infer the correct type

File: src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt [hunk 2]
```diff
@@ -8,11 +9,17 @@ internal class HealthCheckInjector : KoinComponent {
     val dbc: DatabaseConnection by inject()
 }
 
-fun Health.Configuration.liveness() {
-    liveCheck("alive") { true }
-}
+fun Health.Configuration.liveness() = liveCheck("alive") { true }
```
Comment (felix19350) [inline src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt:12]: Shouldn't the liveness check also use the database? I.E. this is the check that is supposed to be used when the application is up and running.

Comment (felix19350) [reply src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt:12]: I'm basing this on the definition of liveness and readyness probes here:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes

File: src/main/kotlin/test/ktortemplate/Application.kt (hunk 1 above)
Comment (rmpt) [reply src/main/kotlin/test/ktortemplate/Application.kt:80]: no, propertyOrNull returns an ApplicationConfigValue, so there's no possible automatic conversion from that to Long.

File: src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt (hunk 2 above)
Comment (rmpt) [reply src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt:12]: not sure about that. Based on the same definition, liveness should check if the system is "dead", so it shouldn't go much deep. On the other hand, readiness should check if the system is ready to serve traffic, so it should go deep and check if everything is in place.
Besides that, what is the point of having the same check in both probes? If both have the same checks, we can use just one endpoint for both liveness and readiness

File: src/main/kotlin/test/ktortemplate/Application.kt (hunk 1 above)
Comment (felix19350) [reply src/main/kotlin/test/ktortemplate/Application.kt:80]: yeah I was just mentioning that you can remove the type `ApplicationConfigValue?` from the variable declaration, as the type system should be able to correctly infer the type.

File: src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt (hunk 2 above)
Comment (felix19350) [reply src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt:12]: Yeah, I get your point but I got a bit of mixed feelings for that in the sense that during operation (where the liveness will be called repeatedly), I want things to warn me when the database goes "down" - i.e. takes so long to respond to the simplest possible query that it is essentially dead.
By not doing that check in the liveness probe, you risk having a liveness probe that says that everything is Okay even when there is no database connection.

Comment (rmpt) [reply src/main/kotlin/test/ktortemplate/core/utils/healthcheck/HealthCheck.kt:12]: isn't readiness prob configurable, just like liveness? Why switch concepts when you can just set a lower interval between readiness endoint calls?

File: .github/workflows/gradle.yml [hunk 3]
```diff
@@ -25,5 +25,11 @@ jobs:
           ${{ runner.os }}-gradle-
     - name: Build with Gradle
       run: ./gradlew build
+    - name: Archive test results
+      uses: actions/upload-artifact@v2
```
Comment (felix19350) [inline .github/workflows/gradle.yml:29]: Might be good to be able to download the test artifacts.
https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts

File: src/test/kotlin/test/ktortemplate/core/utils/healthcheck/TestHealthCheck.kt [hunk 4]
```diff
@@ -23,44 +26,79 @@ class TestHealthCheck : KoinTest {
...
     }
 
     @Test
-    fun `Checking readiness with 500 Internal Server Error`() = testAppWithConfig {
-        with(handleRequest(HttpMethod.Get, "/readiness")) {
-            response.status() `should be equal to` io.ktor.http.HttpStatusCode.OK
-            val result: Map<String, String> = JsonSettings.mapper.readValue(response.content!!)
-            result["database"].toBoolean() `should be equal to` true
+    fun `Checking readiness with 500 Internal Server Error`() {
+        val dbContainer = PgSQLContainerFactory.newInstance()
```
Comment (felix19350) [inline src/test/kotlin/test/ktortemplate/core/utils/healthcheck/TestHealthCheck.kt:52]: I suspect the problem had to do with a timing issue in shutting down the container, so for this test we create a new container that is manually managed

File: src/test/kotlin/test/ktortemplate/core/utils/healthcheck/TestHealthCheck.kt [hunk 5]
```diff
@@ -23,44 +26,79 @@ class TestHealthCheck : KoinTest {
...
 
-        dbContainer.stop()
-        with(handleRequest(HttpMethod.Get, "/readiness")) {
-            response.status() `should be equal to` io.ktor.http.HttpStatusCode.InternalServerError
-            val result: Map<String, String> = JsonSettings.mapper.readValue(response.content!!)
-            result["database"].toBoolean() `should be equal to` false
+            dbContainer.stop()
+            // The wait here is important to avoid timing issues
+            log.info("Stopping container")
+            waitFor(Duration.ofSeconds(3)) {
```
Comment (felix19350) [inline src/test/kotlin/test/ktortemplate/core/utils/healthcheck/TestHealthCheck.kt:67]: Also we wait for the container to be stopped before proceeding.

**Review by felix19350 — APPROVED** [review]

**Pull request merged**